api/src/main/java/org/pmiops/workbench/api/ClusterController.java

```diff
@@ -114,29 +143,61 @@ public Cluster apply(org.pmiops.workbench.notebooks.model.Cluster firecloudClust
       }
       throw ExceptionUtils.convertFirecloudException(e);
     }
+    CdrVersion cdrVersion =
+        workspaceService.getRequired(body.getWorkspaceNamespace(), body.getWorkspaceId()).getCdrVersion();
 
     // For the common case where the notebook cluster matches the workspace
     // namespace, simply name the directory as the workspace ID; else we
     // include the namespace in the directory name to avoid possible conflicts
     // in workspace IDs.
+    String gcsNotebooksDir = "gs://" + fcWorkspace.getBucketName() + "/notebooks";
     String workspacePath = body.getWorkspaceId();
     if (!projectName.equals(body.getWorkspaceNamespace())) {
       workspacePath = body.getWorkspaceNamespace() + ":" + body.getWorkspaceId();
     }
-    String apiDir = String.join("/", "workspaces", workspacePath);
-    String localDir = String.join("/", "~", apiDir);
-    Map<String, String> toLocalize = body.getNotebookNames()
-        .stream()
-        .collect(Collectors.<String, String, String>toMap(
-            name -> localDir + "/" + name,
-            name -> String.join("/", workspaceBucket, "notebooks", name)));
-    // TODO(calbach): Localize a delocalize config JSON file as well, once Leo supports this.
-    toLocalize.put(
-        localDir + "/.all_of_us_config.json",
-        workspaceBucket + "/" + WorkspacesController.CONFIG_FILENAME);
-    notebooksService.localize(projectName, clusterName, toLocalize);
+
+    // Materialize the JSON config files directly via the Jupyter server API.
+    // TODO: It is very inefficient to do 5 serial requests here. Rework this
+    // to utilize "mkdir -p"-like functionality, e.g. via starting with calling
+    // the /localize endpoint.
+    notebooksService.putRootWorkspacesDir(projectName, clusterName);
```

Comment:
What is the behavior of this if it already exists? Since we are calling this each time we localize? Or does this only happen once?

Reply:
Good question. Added a comment to better explain this. If the directories/files already exist these calls are no-ops. Let me know if the comment clears up the rest.